rdrf/rdrf/reports/generator.py

```diff
@@ -0,0 +1,707 @@
+            return self._get_last_user(patient_model, context_model)
+        elif self.field == "timestamp":
+            return patient_model.get_form_timestamp(self.form_model, context_model)
+        elif self.field == "progress":
+            return self._get_form_progress(patient_model, context_model)
+        else:
+            raise Exception("Unknown field: %s" % self.field)
+
+    def _get_form_progress(self, patient_model, context_model):
+        if self.progress_percentage is None:
+            self.progress_percentage = self.form_progress.get_form_progress(self.form_model,
+                                                                            patient_model,
+                                                                            context_model)
+        return self.progress_percentage
+                                                                    
+        
+
+    def _get_last_user(self, patient_model, context_model):
+        # last user to edit the _form_ in this context
+        history = ClinicalData.objects.collection(self.registry_model.code, "history")
+        snapshots = history.find(patient_model, record_type="snapshot")
+        snapshots = sorted([s for s in snapshots], key=attrgetter("pk"), reverse=True)
+        logger.debug("got %s snapshots" % len(snapshots))
+        for snapshot in snapshots:
+            if snapshot.data and "record" in snapshot.data:
+                record = snapshot.data["record"]
+                if "context_id" in record:
+                    if context_model.pk == record["context_id"]:
+                        if "form_name" in snapshot.data:
+                            form_name = snapshot.data["form_name"]
+                            if form_name == self.form_model.name:
+                                if "form_user" in snapshot.data:
+                                    return snapshot.data["form_user"]
+
+
+    def _get_cde_value(self, patient_model, context_model):
+        try:
+            data = get_clinical_data(self.registry_model.code,
+                                     patient_model.pk,
+                                     context_model.pk)
+            
+
+            raw_value = patient_model.get_form_value(self.registry_model.code,
+                                                     self.form_model.name,
+                                                     self.section_model.code,
+                                                     self.cde_model.code,
+                                                     False,
+                                                     context_id=context_model.pk,
+                                                     clinical_data=data)
+
+        except KeyError:
+            # the dynamic data record is empty
+            return None
+
+        value = self.cde_model.get_display_value(raw_value)
+        return fix_display_value(self.cde_model.datatype, value)
+
+
+def pg_uri(db):
```

Comment:
Use `sqlalchemy.engine.url.URL` instead of re-implementing the construction of the URL.